src/Operator/ProductDraftTaxonsOperator.php

```diff
@@ -54,12 +65,46 @@ public function updateTaxonsInProduct(ProductDraftInterface $productDraft, Produ
             $product->setMainTaxon(null);
         }
 
-        $productTaxons = $product->getProductTaxons();
-        foreach ($productTaxons as $productTaxon) {
-            $product->removeProductTaxon($productTaxon);
-            $this->entityManager->remove($productTaxon);
+        $taxonIdsFromProduct = $this->getTaxonIdsForProduct($product);
+
+        $taxonIdsFromProductDraft = $this->getTaxonIdsForProductDraft($productDraft);
+
+        $sharedTaxonIds = array_intersect($taxonIdsFromProduct, $taxonIdsFromProductDraft);
+
+        /** @var ProductTaxon $productTaxon */
+        foreach ($product->getProductTaxons() as $productTaxon) {
+            /** @var TaxonInterface $taxon */
+            $taxon = $productTaxon->getTaxon();
+            if (!in_array($taxon->getId(), $sharedTaxonIds)) {
+                $product->removeProductTaxon($productTaxon);
+                $this->entityManager->remove($productTaxon);
+            }
         }
 
         $this->copyTaxonsToProduct($productDraft, $product);
     }
+
+    private function getTaxonIdsForProduct(ProductInterface $product): array
+    {
+        $taxonIds = [];
+        foreach ($product->getProductTaxons() as $productTaxon) {
+            /** @var TaxonInterface $taxon */
+            $taxon = $productTaxon->getTaxon();
+            $taxonIds[] = $taxon->getId();
+        }
+
+        return $taxonIds;
```

Comment:
What about this? (not tested)
```suggestion
        return array_map(fn($productTaxon) => $productTaxon->getTaxon()->getId(), $product->getProductTaxons());
```
Can be done for `getTaxonIdsForProductDraft` the same way.

Reply:
Nice one liner, would prefer it much more, but had to provide that type hint for TaxonInterface.

Reply:
@MacBalc why is that needed? `public function getTaxon(): ?TaxonInterface;` from the `ProductTaxonInterface` Will always return a `TaxonInterface` or `null` and you don't have a null check here so I don't see any benefit on that.

Reply:
@maxbeckers it is needed due to phpstan checks